File: Leetcode/problem1423.js
function maxScore(cardPoints, k) {
    const n = cardPoints.length;
    let leftsum = 0;
    let rightSum = 0;
    let maxSum = 0;

    for (let i = 0; i <= k - 1; i++) {
        leftsum += cardPoints[i];
        maxSum = leftsum;
    }

    let rightIndex = n - 1;
    for (let i = k - 1; i >= 0; i--) {
        leftsum = leftsum - cardPoints[i];
        rightSum += cardPoints[rightIndex];
        rightIndex--;

        maxSum = Math.max(maxSum, leftsum + rightSum);
    }

    return maxSum;
}

// TC - O(2*k)
// SC - O(1)

// Test Cases
console.log(maxScore([1, 2, 3, 4, 5, 6, 1], 3));

File: Leetcode_CPP/problem1423.cpp
#include <iostream>
#include <stdio.h>
#include <vector>
#include <algorithm>
using namespace std;

// optimal
// time - O(k)
// space - S(1)
int maxScore(vector<int> &cardPoints, int k)
{
    int n = cardPoints.size();
    int maxPoints = 0, left = 0, right = n - 1, taken = 0;
    long sum = 0;

    while (right >= n - k)
    {
        while (taken < k)
        {
            sum += cardPoints[left];
            maxPoints = sum;
            taken++;
            if (taken < k)
                left++;
        }

        sum -= cardPoints[left];

        sum += cardPoints[right];
        if (maxPoints < sum)
            maxPoints = sum;

        taken++;
        left--;
        right--;
    }

    return maxPoints;
}

int maxScore(vector<int> &cardPoints, int k)
{
    int n = cardPoints.size();
    int maxPoints = 0, leftSum = 0, rightSum = 0;

    for (int i = 0; i < k; i++)
    {
        leftSum += cardPoints[i];
    }

    maxPoints = leftSum;

    int right = n - 1;
    for (int i = k - 1; i >= 0; i--)
    {
        rightSum += cardPoints[right];
        leftSum -= cardPoints[i];
        right--;
        maxPoints = max(maxPoints, leftSum + rightSum);
    }

    return maxPoints;
}

int main()
{
    return 0;
}